src/sink.rs

```diff
@@ -0,0 +1,847 @@
+//! Provides a `Sink` interface for publishing to hedwig.
+
+use crate::{Message, PublishBatch, Publisher, Topic, ValidatedMessage};
+use either::Either;
+use futures_util::{ready, sink::Sink, stream::Stream};
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// A sink which ingests messages, validates them with the given validator, then publishes them
+/// with the given publisher.
+///
+/// This is a convenience function which combines [`validator_sink`](validator_sink) and
+/// [`publisher_sink`](publisher_sink)
+pub fn sink<M, P>(
+    validator: M::Validator,
+    publisher: P,
+    batch_size: usize,
+) -> ValidatorSink<M, M::Validator, PublisherSink<P, P::PublishStream>>
+where
+    M: Message,
+    P: Publisher,
+    P::PublishStream: Unpin,
+{
+    // TODO a retry layer could go between these two
+    validator_sink(validator, publisher_sink(publisher, batch_size))
+}
+
+/// A sink which ingests messages, validates them with the given validator, then forwards them to
+/// the given destination sink.
+pub fn validator_sink<M, S>(
```

Comment:
the validator can be just a stream transformer, yes. It could also be just a sink transformer too, via [`with`](https://docs.rs/futures/0.3.13/futures/sink/trait.SinkExt.html#method.with). In a sense the ValidatorSink is just a `With` but as a concrete type instead of a transformer with a closure.

In any case, the retry logic would be optional as another Sink adapter. End users could choose to use it or not, that's why these steps are broken into separate Sinks instead of one big one. I do think it's valuable to provide _some_ kind of retry option here so that users don't have to implement it themselves, but of course they'd still be free to do so if they have more specialized needs.

Reply:
My problem is probably more with the `sink` method trying to give an out-of-box experience, while its full of major caveats.

I think the `sink` function does not pull its weight as written now and should be removed from this PR. With absence of the `sink` function users are implicitly directed to think about how they want to combine the building blocks provided here and how they ought to implement things like retrying.

In that sense I also suspect my concerns about the publisher batching would be mostly resolved if the publisher sink was split into two elements – one which batches, and another which publishes the batches.

Reply:
I suppose I don't agree that the caveats are so major that having a simplistic `sink` available is harmful. The docs label it a convenience function, its constituent parts are linked, and the docs on the parts should hopefully describe them enough.

This isn't really a building block kind of library anyway, `keeping things pretty simple and straight forward` is on the front page. From a principled perspective, yes having some sink that batches and then a sink that knows how to publish batches would be more flexible and provide more choice -- though other questions like the user latency vs impl throughput are still an issue. But would *we* actually make use of that? This is in a state where our current project's needs are practically met, I'd prefer that we approach this pragmatically

Reply:
> I'd prefer that we approach this pragmatically
yes please

Reply:
After some discussion in slack, I've removed the `sink` function so that there's less concern over pushing users into unexpected choices